server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2AuthenticationParametersImpl.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.authentication;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.io.UnsupportedEncodingException;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import static java.net.URLDecoder.decode;
+import static java.net.URLEncoder.encode;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.sonar.server.authentication.Cookies.findCookie;
+import static org.sonar.server.authentication.Cookies.newCookieBuilder;
+
+public class OAuth2AuthenticationParametersImpl implements OAuth2AuthenticationParameters {
+
+  private static final String AUTHENTICATION_COOKIE_NAME = "AUTHENTICATION_COOKIE";
+
+  /**
+   * The HTTP parameter that contains the path where the user should be redirect to.
+   * Please note that the web context is included.
+   */
+  private static final String RETURN_TO_PARAMETER = "return_to";
+
+  private static final Type JSON_MAP_TYPE = new TypeToken<HashMap<String, String>>() {
+  }.getType();
+
+  @Override
+  public void init(HttpServletRequest request, HttpServletResponse response) {
+    String returnTo = request.getParameter(RETURN_TO_PARAMETER);
+    response.addCookie(newCookieBuilder(request)
+      .setName(AUTHENTICATION_COOKIE_NAME)
+      .setValue(toJson(ImmutableMap.of(RETURN_TO_PARAMETER, returnTo)))
+      .setHttpOnly(true)
+      .setExpiry(-1)
```

Comment:
It would be cleaner to put an expiration date, for instance 5 minutes later. Note that it's the opportunity to document the class `CookieBuilder`. I don't know what's the parameter (a date ? a duration ? which unit of time ?)

Reply:
Indeed, I also thing that we should put an expiration date, and to add JavaDoc.